### ISSUE TITLE
Webcam callback hooks execute exponentially (potential memory leak).

### ISSUE THREAD

**Issue by fyrye** [jhuckaby/webcamjs#87]
Seems to happen in either Chrome or FireFox and with either using `on.('uploadComplete')`, or `upload(data, '/path/to/save', callback)`.

Essentially running snap 3 times will log 
`a1, b1, a2, b2, b3, a3, b4, b5, b6`
instead of 
`a1, b1, a2, b2, a3, b3`

```javascript
var a = 1, b = 1;
Webcam.snap(function(data_uri, canvas, context) {
    console.log('a' + a++);
    Webcam.on('uploadComplete', function(code, text) {
        console.log('b' + b++);
    });
    Webcam.upload(data_uri, '/path/to/save');
});
```
or 
```javascript
Webcam.snap(function(data_uri, canvas, context) {
    console.log('a' + a++);
    Webcam.upload(data_uri, '/path/to/save', function(code, text) {
        console.log('b' + b++);
    });
});
```

Additionally tried calling `Webcam.reset(); Webcam.attach();` in the callback with the same results.

Issue appears to be related to the `Webcam.on` and `Webcam.dispatch` methods, though appears as it is the desired functionality.

```javascript
	on: function(name, callback) {
		// set callback hook
		name = name.replace(/^on/i, '').toLowerCase();
		if (!this.hooks[name]) this.hooks[name] = [];
		this.hooks[name].push( callback );
	},
```
Should be
```javascript
    on: function(name, callback) {
        // set callback hook
        name = name.replace(/^on/i, '').toLowerCase();
        if (!this.hooks[name]) {
            this.hooks[name] = [];
            this.hooks[name].push(callback);
        }
    },
```
Or
```javascript
  on: function(name, callback) {
        // set callback hook
        name = name.replace(/^on/i, '').toLowerCase();
        if (!this.hooks[name]) {
            this.hooks[name] = callback;
        }
    },
```
Then change off and dispatch to reflect the change.

Or to keep appending to the hooks array for General Usage, documents should note the usage of `.on` and `.off` for expected results.
```javascript
Webcam.snap(function(data_uri, canvas, context) {
    console.log('a' + a++);
    var callback = function(code, text){
        console.log('b' + b++);
        Webcam.off('uploadComplete', callback);
    };
    Webcam.upload(data_uri, '/path/to/save', callback);
});
```

**Comment by jhuckaby** [comment]
This is as designed behavior.  Each call to `on()` is supposed to append the handler to an array, so multiple listeners can listen in on the same event.  Your example which uses `off()` is correct usage.  I will update the docs to explain this behavior.  Thanks for the detailed bug report!

**Comment by jhuckaby** [comment]
Ah ha, you're absolutely right regarding passing an inline callback to `Webcam.upload()`.  This totally accumulates the callback onto the event array, which is a bad situation, and not intended.  I just fixed this one, and updated the README as well:

https://github.com/jhuckaby/webcamjs/commit/5f91ae0b380ae4193cbd7868a9c927e7d77249c2

The usage of `on()` and `off()` is as designed, however, and I added an explanation of that behavior in the docs.

Thanks for pointing this out!
